python/triqs/lattice/utils.py

```diff
@@ -246,3 +248,171 @@ def TB_from_pythTB(ptb):
+    # hops contains details about hopping of electrons such as orbital
+    # and hopping amplitude
+    if TBL.units.shape == (2, 2):
+        TBL_units = np.eye(3)
+        TBL_units[:2, :2] = TBL.units
+        TBL_hops = {key + (0,): val for key, val in TBL.hoppings.items()}
+    elif TBL.units.shape == (3,3):
+        TBL_units = TBL.units
+        TBL_hops = TBL.hoppings
+    # raises error for when the dimensions of the tb object is neither 2D nor 3D
+    else:
+        raise ValueError("This format of the tight-binding model is not implemented for this function.")
+   
+    # number of orbitals involved in the unit cell
+    num_orb = TBL.n_orbitals
+
+    # maximum hopping distances of electrons in each direction
+    max_x, max_y, max_z = list(np.max(np.array(list(TBL_hops.keys())), axis = 0))
+
+    # number of cells involved in the hopping of electrons in each direction
+    num_cells_x, num_cells_y, num_cells_z = [2 * max_coord + 1 for max_coord in [max_x, max_y, max_z]]
+    
+    # real-space Hamiltonian
+    Hrij = np.zeros((num_cells_x, num_cells_y, num_cells_z, num_orb, num_orb), dtype = sp.exp)
+
+    # looping through hopping parameters of electrons involved in inter-orbital hoppings
+    for key, hopping in TBL_hops.items():
+        rx, ry, rz = key
+        hopping = np.around(hopping, precision)
+        Hrij[rx + max_x, ry + max_y, rz + max_z] = hopping
+
+    # basis of the exponential term in calculation of Hk
+    Hexp = np.empty_like(Hrij, dtype = sp.exp)
+
+    # perform Fourier transform
+    for xi, yi, zi in itp(range(num_cells_x), range(num_cells_y), range(num_cells_z)):
+        coefficients = np.array([xi - max_x, yi - max_y, zi - max_z])
+        r = lattice.dot(coefficients)
+        eikr = sp.exp(-I * r)
+        Hexp[xi, yi, zi, :, :] = eikr
+
+    # summation over all real space axes
+    Hk = np.sum(Hrij * Hexp, axis = (0, 1, 2))
+    
+    # rewriting exponential terms in Hamiltonian expression in terms of cosine
+    for i, j in itp(range(num_orb), repeat = 2):
+        Hk[i, j] = Hk[i, j].rewrite(sp.cos)
+
+    # dealing with the numerical Hamiltonian
+
+    # convert to SymPy matrix to use substitutions method available in SymPy
+    Hk_numerical = sp.Matrix(Hk)
+
+    # obtaining individual displacement vectors
+    a1 = np.around(TBL_units[0], precision)
+    a2 = np.around(TBL_units[1], precision)
+    a3 = np.around(TBL_units[2], precision)
+
+    # numerical dot products between unit vectors
+    # and momentum space matrix
+    a1k_numerical = a1.dot(k_space_matrix)[0]
+    a2k_numerical = a2.dot(k_space_matrix)[0]
+    a3k_numerical = a3.dot(k_space_matrix)[0]
+    
+    # performing numerical dot product substitutions
+    Hk_numerical = Hk_numerical.subs(a1k, a1k_numerical)
+    Hk_numerical = Hk_numerical.subs(a2k, a2k_numerical)
+    Hk_numerical = Hk_numerical.subs(a3k, a3k_numerical)
+
+    # converting numerical Hamiltonian to NumPy array
+    Hk_numerical = np.array(Hk_numerical)
```

Comment:
```suggestion
    return np.array(Hk_numerical)
```

Reply:
I would like to first check whether the user wants to return an analytical or numerical form before returning anything. Moreover, returning `np.array(Hk_numerical)` that early on will cause the function to "exit" before even initiating the checks afterward.

Reply:
Since the checks I instituted, `_has_complex_exponential_sympy(matrix)` and `_is_hermitian_sympy(matrix)`, depend on `Hk_numerical` being a NumPy array, I first convert it from a SymPy matrix to a NumPy array and then do the checks.